### PR TITLE
NEXT-12791 - [NEW] Add responsive behavior

### PR DESCRIPTION
**Pull request by NiklasLimberg**
h2. What should be done?

Create a new article on our [GitBook instance|https://app.gitbook.com/@shopware/s/shopware/] which explains how to add responsive behavior to the administration (e.g. a custom module).

This article should mention:
- The prerequisite, a working plugin (Refer to the plugin base guide)
- Every other prerequisite you figure out during writing the guide (e.g. a subscriber, knowing how to create a service, a controller, etc.)
- A short code example, including an explanation, on how to do it

Category: Extensions > Plugins > Administration

h2. Are there already guides in the previous documentation for this?
Unfortunately, No. But please have a look at our [example guide|https://app.gitbook.com/@shopware/s/shopware/guides/plugins/plugins/plugin-base-guide](e.g. note the PLACEHOLDERs for cross-references, that do not exist yet) and the [Writing guidelines|https://app.gitbook.com/@shopware/s/shopware/resources/guidelines/documentation/writing].

For more information, ask me, [~p.stahl].

### CODE REVIEW

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Adding responsive behavior
+
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:2]:
```suggestion

## Overview
```

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+The Shopware 6 Administration provides two ways of adding classes to elements based on their size, the device helper and the `v-responsive` directive.
+Alternatively you can use `sccs` media queries to make your plugin responsive.
+Learn how to use `sccs` here, [Add custom styles](./add-custom-styles.md).
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:5]: `page-ref` nutzen?

```suggestion
Learn how to use `scss` here:

{% page-ref page="./add-custom-styles.md" %}
```

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+## The device helper
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:7]:
```suggestion
## DeviceHelper
```

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+const listener = function (ev) {
+    // do something on resize with the event, like passing  
+};
+const scope = this;
+const component = 'sw-basic-example';
+
+this.$device.onResize({ listener, scope, component });
+this.$device.removeResizeListener(component);
+```
+It also provides many helper functions to get for example the screen dimensions.
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:26]:
```suggestion
It also provides many helper functions e.g. to get the screen dimensions.
```

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+Although there are many more as seen below:
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:27]: Can or should we have a reference article about that? //CC: @PaddyS

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+## v-responsive directive
+
+The `v-responsive` directive can be used to dynamically apply classes based on an elements dimensions.
+
+```html
+<input v-responsive="{ 'is--compact': el => el.width <= 1620, timeout: 200 }">
+```
+
+Explanation:
+- Apply class (in this case: 'is--compact') when the width of the element is smaller than 1620px.
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:49]:
```suggestion
- Apply class (in this case: `is--compact`) when the width of the element is smaller than 1620px.
```

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+- timeout: Sets the duration on how much the throttle should wait.
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:50]:
```suggestion
- `timeout`: Sets the duration on how much the throttle should wait.
```

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+## v-responsive directive
+
+The `v-responsive` directive can be used to dynamically apply classes based on an elements dimensions.
+
+```html
+<input v-responsive="{ 'is--compact': el => el.width <= 1620, timeout: 200 }">
+```
+
+Explanation:
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:48]:
```suggestion
Let's do a small explanation of this directive:
```

**Review by leichteckig — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/administration/adding-responsive-behavior.md (hunk 5 above)
Comment (NiklasLimberg) [reply guides/plugins/plugins/administration/adding-responsive-behavior.md:27]: I really don't know what this refernce would be about

Comment (leichteckig) [reply guides/plugins/plugins/administration/adding-responsive-behavior.md:30]: I would explain the directives you list right after this line: 

```javascript
this.$device.getViewportWidth();
this.$device.getViewportHeight()
this.$device.getDevicePixelRatio()
this.$device.getScreenWidth()
this.$device.getScreenHeight()
this.$device.getScreenOrientation()
```

**Review by leichteckig — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/administration/adding-responsive-behavior.md (hunk 5 above)
Comment (PaddyS) [reply guides/plugins/plugins/administration/adding-responsive-behavior.md:30]: I like.

**Review by raknison — APPROVED** [review]

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+this.$device.onResize({ listener, scope, component });
+this.$device.removeResizeListener(component);
+```
+It also provides many helper functions e.g. to get the screen dimensions.
+Although there are many more as seen below:
+
+| Function                               | Description                 |
+|----------------------------------------|-----------------------------|
+| `this.$device.getViewportWidth();`     | gets the viewport width     |
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:34]: Please use capital letters at the beginning of the sentences in the second column

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+| `this.$device.getViewportHeight();`    | gets the viewport height    |
+| `this.$device.getDevicePixelRatio();`  | gets the device pixel ratio |
+| `this.$device.getScreenWidth();`       | gets the screen width       |
+| `this.$device.getScreenHeight();`      | gets screen height          |
+| `this.$device.getScreenOrientation();` | gets the screen orientation |
+
+
+## v-responsive directive
+
+The `v-responsive` directive can be used to dynamically apply classes based on an elements dimensions.
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:44]:
```suggestion
The `v-responsive` directive can be used to dynamically apply classes based on an element's dimensions.
```

**Review by leichteckig — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+The Shopware 6 Administration provides two ways of adding classes to elements based on their size, the device helper and the `v-responsive` directive.
+Alternatively you can use `sccs` media queries to make your plugin responsive.
+Learn how to use `scss` here:
+
+{% page-ref page="./add-custom-styles.md" %}
+
+## DeviceHelper
+
+The DeviceHelper provides methods to get device and browser information like the current viewport size.
+The helper methods can be accessed with "this.$device" in every Vue component,
```
Comment (leichteckig) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:13]:
```suggestion
The helper methods can be accessed with `this.$device` in every Vue component,
```

**Review by leichteckig — CHANGES_REQUESTED** [review]
Just one more small thing :)

**Review by leichteckig — APPROVED** [review]

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
+# Adding responsive behavior
+
+## Overview
+The Shopware 6 Administration provides two ways of adding classes to elements based on their size, the device helper and the `v-responsive` directive.
+Alternatively you can use `sccs` media queries to make your plugin responsive.
```
Comment (JoshuaBehrens) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:5]: `sccs` -> `scss`

How are scss media queries different css media queries?

**Review by JoshuaBehrens — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/administration/adding-responsive-behavior.md (hunk 12 above)
Comment (lernhart) [reply guides/plugins/plugins/administration/adding-responsive-behavior.md:5]: There is no difference at all. I'd suggest you refer to `CSS` queries here

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+{% page-ref page="./add-custom-styles.md" %}
+
+## DeviceHelper
+
+The DeviceHelper provides methods to get device and browser information like the current viewport size.
+The helper methods can be accessed with `this.$device` in every Vue component,
+since it is bound to the Vue prototype.
+
+It provides functionality to run a function on resize to for example add classes to a element.
```
Comment (lernhart) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:16]: Please revisit this sentence. That's not proper English and I find it hard to suggest anything here, as I do not understand, what you want to say here

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+The example below shows you how to use the `$device.onResize` helper.
+
+```javascript
+const listener = function (ev) {
+    // do something on resize with the event, like passing
```
Comment (lernhart) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:21]: Passing ... ?

File: guides/plugins/plugins/administration/adding-responsive-behavior.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+```javascript
+const listener = function (ev) {
+    // do something on resize with the event, like passing  
+};
+const scope = this;
+const component = 'sw-basic-example';
+
+this.$device.onResize({ listener, scope, component });
+this.$device.removeResizeListener(component);
```
Comment (lernhart) [inline guides/plugins/plugins/administration/adding-responsive-behavior.md:27]: Why do you remove the listener afterwards ? Please document this

**Review by lernhart — CHANGES_REQUESTED** [review]

**Review by lernhart — APPROVED** [review]

**Comment by JoshuaBehrens** [comment]
I approve :D

**Pull request merged**